### PR TITLE
perf(dbapi): avoid copy config in hot path

### PR DESCRIPTION
**Pull request by majorgreys**
Rather than copying and calling a merge on dbapi2 config with the config for the dbapi2 compliant library, use a proxy configuration object that looks up the integration-specific configuration, and falls back on the dbapi2 configuration.

### CODE REVIEW

**Comment by brettlangdon** [comment]
do you have before/after performance numbers?

**Comment by majorgreys** [comment]
> do you have before/after performance numbers?

@P403n1x87 will be posting those

**Comment by P403n1x87** [comment]
> > do you have before/after performance numbers?
> 
> @P403n1x87 will be posting those

Unfortunately, this is still not enough, as it seems that many cursor objects are being created, which still trigger the `copy` on the dbapi2 configuration 🙁 . With this change I can still see the same ~3% overhead, which could be removed almost entirely.

File: ddtrace/contrib/dbapi/__init__.py [hunk 1]
```diff
@@ -164,7 +165,7 @@ def fetchmany(self, *args, **kwargs):
         )
 
 
-def _get_config(new_cfg):
+def _override_dbapi2_config(new_cfg):
```
Comment (P403n1x87) [inline ddtrace/contrib/dbapi/__init__.py:168]: An alternative interim change, before the dbapi2 config is removed for good, could be
~~~ python
class OverrideAttrDict(wrapt.ObjectProxy):
    __slots__ = ("override", "base")
    sentinel = object()

    def __init__(self, override, base):
        super(OverrideAttrDict, self).__init__(override)
        self.override = override
        self.base = base

    def __getattr__(self, name):
        value = self.override.get(name, self.sentinel)
        return getattr(self.base, name) if value == self.sentinel else value

    def __getitem__(self, name):
        value = self.override.get(name, self.sentinel)
        return self.base.__getitem__(name) if value == self.sentinel else value


def _override_dbapi2_config(new_cfg):
    # Need to backwards support the dbapi2 config entry
    # but give precedence to the given config.

    return OverrideAttrDict(new_cfg, config.dbapi2)
~~~
Profiling data show that the overhead goes down to 0.1%

Comment (majorgreys) [reply ddtrace/contrib/dbapi/__init__.py:187]: Thanks for the tip @P403n1x87! I've added it in.

Comment (P403n1x87) [reply ddtrace/contrib/dbapi/__init__.py:187]: @brettlangdon with @majorgreys' latest change, the total overhead of the new `_override_dbapi2_config` is around 0.3% (on average, per-request); with the old `_get_config`, the total overhead is around 6.3%.

File: ddtrace/contrib/dbapi/__init__.py [hunk 2]
```diff
@@ -164,12 +165,30 @@ def fetchmany(self, *args, **kwargs):
...
+
+    def __getattr__(self, name):
+        value = self.override.get(name, self.sentinel)
+        return getattr(self.base, name) if value == self.sentinel else value
+
+    def __getitem__(self, name):
+        value = self.override.get(name, self.sentinel)
+        return self.base.__getitem__(name) if value == self.sentinel else value
+
+
```
Comment (P403n1x87) [inline ddtrace/contrib/dbapi/__init__.py:186]: Need to add

~~~ python

    def __bool__(self):
        return bool(self.override or self.base)

    def __contains__(self, item):
        return item in self.override or item in self.base
~~~

File: ddtrace/contrib/dbapi/__init__.py [hunk 3]
```diff
@@ -164,12 +165,30 @@ def fetchmany(self, *args, **kwargs):
...
-def _get_config(new_cfg):
+# TODO: Remove once config.dbapi2 has been removed
+class _OverrideAttrDict(wrapt.ObjectProxy):
+    __slots__ = ("override", "base")
+    sentinel = object()
+
+    def __init__(self, override, base):
+        super(_OverrideAttrDict, self).__init__(override)
+        self.override = override
+        self.base = base
```
Comment (P403n1x87) [inline ddtrace/contrib/dbapi/__init__.py:176]:
```suggestion
        self.override = override or AttrDict()
        self.base = base or AttrDict()
        super(_OverrideAttrDict, self).__init__(self.override)
```

File: ddtrace/contrib/dbapi/__init__.py (hunk 2 above)
Comment (P403n1x87) [reply ddtrace/contrib/dbapi/__init__.py:190]: @majorgreys  I think the `__bool__` must be implemented too

Comment (majorgreys) [reply ddtrace/contrib/dbapi/__init__.py:192]: Wondering what your reasoning was for this. Should we still add it in?

Comment (P403n1x87) [reply ddtrace/contrib/dbapi/__init__.py:192]: It made the test for dbapi pass with my experiments so I thought it was needed. It looks like all tests are now passing without it so there's no need for it I suppose.

**Review by P403n1x87 — APPROVED** [review]

**Review by brettlangdon — APPROVED** [review]

**Pull request merged**